File: 3090-maximum-length-substring-with-two-occurrences/3090-maximum-length-substring-with-two-occurrences.java
class Solution {

    public int maximumLengthSubstring(String s) {
        int maxLengthOfSubstring = 0;
        char[] occurrences = new char[26];

        for (int rightPoint = 0, leftPoint = 0; rightPoint < s.length(); rightPoint++) {
            char rightCharacter = s.charAt(rightPoint);
            occurrences[rightCharacter - 'a']++;

            if (occurrences[rightCharacter - 'a'] <= 2) {
                maxLengthOfSubstring = Math.max(maxLengthOfSubstring, rightPoint - leftPoint + 1);
                continue;
            }

            while (occurrences[rightCharacter - 'a'] > 2) {
                occurrences[s.charAt(leftPoint) - 'a']--;
                leftPoint++;
            }
        }

        return maxLengthOfSubstring;
    }
}

File: 3090-maximum-length-substring-with-two-occurrences/3090-maximum-length-substring-with-two-occurrences.js
/**
 * @param {string} s
 * @return {number}
 */
var maximumLengthSubstring = function(s) {
    const freq = new Array(26).fill(0);

    let left = 0;

    let ans = 0;

    for (let right = 0; right < s.length; right++) {
        const index = s.charCodeAt(right) - 97;
        freq[index]++;

        while (freq[index] > 2) {
            freq[s.charCodeAt(left) - 97]--;

            left++;
        }

        ans = Math.max(ans, right - left + 1);
    }

    return ans;
};